app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt

```diff
@@ -0,0 +1,104 @@
+package de.westnordost.streetcomplete.controls
+
+import android.net.ConnectivityManager
+import android.os.Bundle
+import android.view.View
+import android.widget.Toast
+import androidx.appcompat.app.AlertDialog
+import androidx.core.content.getSystemService
+import androidx.fragment.app.Fragment
+import de.westnordost.osmapi.map.data.BoundingBox
+import de.westnordost.streetcomplete.ApplicationConstants
+import de.westnordost.streetcomplete.Injector
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.download.DownloadController
+import de.westnordost.streetcomplete.ktx.toast
+import de.westnordost.streetcomplete.map.MainFragment
+import de.westnordost.streetcomplete.util.area
+import de.westnordost.streetcomplete.util.asBoundingBoxOfEnclosingTiles
+import de.westnordost.streetcomplete.util.enclosingBoundingBox
+import kotlinx.android.synthetic.main.fragment_main_menu_button.view.*
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import javax.inject.Inject
+import kotlin.math.PI
+import kotlin.math.sqrt
+
+/** Fragment that shows the main menu button and manages its logic */
+class MainMenuButtonFragment : Fragment(R.layout.fragment_main_menu_button),
+    CoroutineScope by CoroutineScope(Dispatchers.Main) {
+
+    @Inject internal lateinit var downloadController: DownloadController
+
+    /* --------------------------------------- Lifecycle ---------------------------------------- */
+
+    init {
+        Injector.applicationComponent.inject(this)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        view.mainMenuButton.setOnClickListener { onClickMainMenu() }
+    }
+
+    /* ------------------------------------------------------------------------------------------ */
+
+    internal fun onClickMainMenu() {
+        context?.let { MainMenuDialog(it, this::onClickDownload).show() }
+    }
+
+    /* ------------------------------------ Download Button  ------------------------------------ */
+
+    private fun onClickDownload() {
+        if (isConnected()) downloadDisplayedArea()
+        else context?.toast(R.string.offline)
+    }
+
+    private fun isConnected(): Boolean {
+        val connectivityManager = context?.getSystemService<ConnectivityManager>()
+        val activeNetworkInfo = connectivityManager?.activeNetworkInfo
+        return activeNetworkInfo != null && activeNetworkInfo.isConnected
+    }
+
+    private fun downloadDisplayedArea() {
+        val displayArea = (requireParentFragment() as MainFragment).mapFragment?.getDisplayedArea()
```

Comment:
This strongly couples this fragment to the MainFragment.

The ususal pattern used in this app for fragments to communicate with their parent is to declare 
```kotlin
    interface Listener {
        fun getDownloadArea(): BoundingBox
    }
    private val listener: Listener? get() = parentFragment as? Listener ?: activity as? Listener
```
, then change the code here to
```kotlin
val displayArea = listener?.getDownloadArea() ?: return
```

This way, the parent could be an activity or fragment, anything that implements that interface.